src/track/trackfile.h

```diff
@@ -0,0 +1,104 @@
+// path/location, URL, and URI representations.
+//
+// All file-related track properties are snapshots from the provided
+// QFileInfo. Obtaining them might involve accessing the file system
+// and should be used consciously! The QFileInfo class does some
+// caching behind the scenes.
+//
+// Please note that the canonical location of QFileInfo may change at
+// any time, when the underlying file system structure is modified.
+// It becomes empty if the file is deleted.
+//
+// Copying an instance of this class is thread-safe, because the
+// underlying QFileInfo is implicitly shared.
+class TrackFile {
+  public:
+    // For backward-compatibility the QString single argument
+    // constructor has not been declared as "explicit". It is
+    // also not strictly necessary and might be removed at some
+    // point to prevent unintended type conversions.
+    /*non-explicit*/ TrackFile(
+            const QString& filePath)
+            : m_fileInfo(filePath) {
+    }
+    explicit TrackFile(
+            QFileInfo fileInfo = QFileInfo())
+            : m_fileInfo(std::move(fileInfo)) {
+    }
+    explicit TrackFile(
+            const QDir& dir,
+            const QString& file = QString())
+            : m_fileInfo(QFileInfo(dir, file)) {
+    }
+
+    operator QFileInfo&() {
+        return m_fileInfo;
+    }
+    operator const QFileInfo&() const {
+        return m_fileInfo;
+    }
+
+    // Local file representation
+    QString location() const {
+        return m_fileInfo.absoluteFilePath();
+    }
+    QString canonicalLocation() const {
+        return m_fileInfo.canonicalFilePath();
+    }
+    QString directory() const {
+        return m_fileInfo.absolutePath();
+    }
+    QString baseName() const {
+        return m_fileInfo.baseName();
+    }
+    QString fileName() const {
+        return m_fileInfo.fileName();
+    }
+    qint64 fileSize() const {
+        return m_fileInfo.size();
+    }
+    QDateTime fileCreated() const {
+        return m_fileInfo.created();
+    }
+    QDateTime fileLastModified() const {
+        return m_fileInfo.lastModified();
+    }
+
+    // Check if the file actually exists on the file system.
+    bool checkFileExists() const {
+        return QFile::exists(location());
+    }
+
+    // Refresh the canonical location if it is still empty, i.e. if
+    // the file may have re-appeared after mounting the corresponding
+    // drive while Mixxx is already running.
+    //
+    // We ignore the case when the user changes a symbolic link to
+    // point a file to an other location, since this is a user action.
+    // We also don't care if a file disappears while Mixxx is running.
+    // Opening a non-existent file is already handled and doesn't cause
+    // any malfunction.
+    QString freshCanonicalLocation();
+
+    // Portable URL representation
+    QUrl locationUrl() const;
+    QUrl canonicalLocationUrl() const;
+
+    // String-encoded URL representation
+    QString locationUri() const;
```

Comment:
I still can't get used to the Uri/Url thing. A uri is not necessarily a string, a uri is a sperset of urls. But the caller is probably looking for a function returning a string. 
Is there such a uri = string thing association somewhere extern?
How about we useing an overload or locationUrlString(),

Reply:
We need a normalized, percent encoded string. If you have a look at QUrl there are many ways to encode the contents as a string. The plain "string" semantic is not sufficient. The string is not needed for display purposes, but for platform-independent, portable identification.

Reply:
In the context of a track the file URL is it's URI. In general the URI may contain other information than that. That's the reason why those functions are attached to `TrackFile` which provides the context.

Reply:
Even tough the function is not use yet, I think I understand the purpose. It is just that the name is not descriptive enough for me. 

I have just found that:
https://tools.ietf.org/html/rfc3986#section-1.2.1
https://tools.ietf.org/html/rfc3986#section-2.4
> Once produced, a URI is always in its percent-encoded form.
  
This clarifies the situation for me. 

Now it is also clear why QUrl::toEncoded() returns a QByteArray. 

I guess that many other contributors stumble in the same way. So we need at least a comment that explains this. Why did we convert the uri to a String?  Maybe it is even more clear if we keep the QByteArray. (appart from the CPU saver)

Reply:
Later the URI needs to be encoded as a JSON string. A plain QByteArray is a bad choice, because you lose the information that it is an UTF-8 encoded string. At this point a wrapper is needed.

Reply:
According to the rfc a URI is pure ascii.

Reply:
US-ASCII